src/Build.UnitTests/BackEnd/ThreadSafeTaskRegistry_Tests.cs

```diff
@@ -0,0 +1,153 @@
+        /// </summary>
+        [Fact]
+        public void RoutingUsesDefaultBehaviorWhenMultiThreadedExecutionDisabled()
+        {
+            var buildParameters = new BuildParameters();
+            buildParameters.MultiThreaded = false;
+            buildParameters.IsOutOfProc = false;
+
+            var componentHost = new MockHost(buildParameters);
+            var requestEntry = CreateMockBuildRequestEntry();
+            var taskHost = new TaskHost(componentHost, requestEntry, new MockElementLocation("test.proj"), null);
+
+            // Should use default behavior (false in this case)
+            taskHost.ShouldTaskExecuteOutOfProc("Message").ShouldBeFalse();
+            taskHost.ShouldTaskExecuteOutOfProc("UnknownTask").ShouldBeFalse();
+        }
+
+        /// <summary>
+        /// Verify that when already out-of-process, tasks always execute out-of-process.
+        /// </summary>
+        [Fact]
+        public void TasksExecuteOutOfProcessWhenAlreadyOutOfProcess()
+        {
+            var buildParameters = new BuildParameters();
+            buildParameters.MultiThreaded = true;
+            buildParameters.IsOutOfProc = true;
+
+            var componentHost = new MockHost(buildParameters);
+            var requestEntry = CreateMockBuildRequestEntry();
+            var taskHost = new TaskHost(componentHost, requestEntry, new MockElementLocation("test.proj"), null);
+
+            // Should always be out-of-process when already out-of-process
+            taskHost.ShouldTaskExecuteOutOfProc("Message").ShouldBeTrue();
+            taskHost.ShouldTaskExecuteOutOfProc("UnknownTask").ShouldBeTrue();
+        }
+
+        /// <summary>
+        /// Verify that thread-safe tasks execute in-process when multi-threaded execution is enabled.
+        /// </summary>
+        [Fact]
+        public void ThreadSafeTasksExecuteInProcessWhenMultiThreadedExecutionEnabled()
+        {
+            var buildParameters = new BuildParameters();
+            buildParameters.MultiThreaded = true;
+            buildParameters.IsOutOfProc = false;
+
+            var componentHost = new MockHost(buildParameters);
+            var requestEntry = CreateMockBuildRequestEntry();
+            var taskHost = new TaskHost(componentHost, requestEntry, new MockElementLocation("test.proj"), null);
+
+            // Thread-safe tasks should execute in-process (return false for out-of-process)
+            taskHost.ShouldTaskExecuteOutOfProc("Message").ShouldBeFalse();
+            taskHost.ShouldTaskExecuteOutOfProc("Warning").ShouldBeFalse();
+            taskHost.ShouldTaskExecuteOutOfProc("Error").ShouldBeFalse();
+        }
+
+        /// <summary>
+        /// Verify that non-thread-safe tasks execute out-of-process when multi-threaded execution is enabled.
+        /// </summary>
+        [Fact]
+        public void NonThreadSafeTasksExecuteOutOfProcessWhenMultiThreadedExecutionEnabled()
```

Comment:
I'd like to see tests that prove a safe task (Message) actually runs in another process. And that an unsafe one (Copy) doesn't.